osu.Game/Beatmaps/BeatmapManager.cs

```diff
@@ -112,29 +113,57 @@ public WorkingBeatmap CreateNew(RulesetInfo ruleset, APIUser user)
         /// The new difficulty will be backed by a <see cref="BeatmapInfo"/> model
         /// and represented by the returned <see cref="WorkingBeatmap"/>.
         /// </summary>
-        public virtual WorkingBeatmap CreateNewBlankDifficulty(BeatmapSetInfo beatmapSetInfo, RulesetInfo rulesetInfo)
+        public virtual WorkingBeatmap CreateNewDifficulty(NewDifficultyCreationParameters creationParameters)
         {
-            // fetch one of the existing difficulties to copy timing points and metadata from,
-            // so that the user doesn't have to fill all of that out again.
-            // this silently assumes that all difficulties have the same timing points and metadata,
-            // but cases where this isn't true seem rather rare / pathological.
-            var referenceBeatmap = GetWorkingBeatmap(beatmapSetInfo.Beatmaps.First());
+            var referenceBeatmap = creationParameters.ReferenceBeatmap;
+            var targetBeatmapSet = creationParameters.BeatmapSet;
 
-            var newBeatmapInfo = new BeatmapInfo(rulesetInfo, new BeatmapDifficulty(), referenceBeatmap.Metadata.DeepClone());
+            BeatmapInfo newBeatmapInfo;
+            IBeatmap newBeatmap;
+
+            if (creationParameters.CreateBlank)
+            {
+                newBeatmapInfo = new BeatmapInfo(creationParameters.Ruleset, new BeatmapDifficulty(), referenceBeatmap.Metadata.DeepClone());
+                newBeatmap = new Beatmap { BeatmapInfo = newBeatmapInfo };
+                foreach (var timingPoint in referenceBeatmap.ControlPointInfo.TimingPoints)
+                    newBeatmap.ControlPointInfo.Add(timingPoint.Time, timingPoint.DeepClone());
+            }
+            else
+            {
+                newBeatmap = referenceBeatmap.Clone();
```

Comment:
I'm not sure if it matters at this point, but this is doing a `MemberwiseClone` which means things like `ControlPointInfo` are going to be shared with the existing beatmap. Bringing it up since this has come up in conversation before.

Reply:
I was aware of this, and relying on the working beatmap cache being invalidated for the entire set to not cause problems. Dunno, copying is still an area that is spotty in many places.